metric_learn/scml.py

```diff
@@ -0,0 +1,647 @@
+                       max_iter=max_iter, verbose=verbose,
+                       preprocessor=preprocessor,
+                       random_state=random_state)
+
+  def fit(self, X, y):
+    """Create constraints from labels and learn the SCML model.
+
+    Parameters
+    ----------
+    X : (n x d) matrix
+        Input data, where each row corresponds to a single instance.
+
+    y : (n) array-like
+        Data labels.
+
+    Returns
+    -------
+    self : object
+      Returns the instance.
+    """
+    X, y = self._prepare_inputs(X, y, ensure_min_samples=2)
+
+    basis, n_basis = self._initialize_basis_supervised(X, y)
+
+    constraints = Constraints(y)
+    triplets = constraints.generate_knntriplets(X, self.k_genuine,
+                                                self.k_impostor)
+
+    triplets = X[triplets]
+
+    return self._fit(triplets, basis, n_basis)
+
+  def _initialize_basis_supervised(self, X, y):
+    """ Constructs the basis set following one of the supervised options in
+    case one is selected.
+    """
+
+    if self.basis == 'LDA':
+      basis, n_basis = self._generate_bases_LDA(X, y)
+    else:
+      basis, n_basis = None, None
+
+    return basis, n_basis
+
+  def _generate_bases_LDA(self, X, y):
+    """ Generates bases for the 'LDA' option.
+
+    The basis set is constructed using Linear Discriminant Analysis of
+    significant local regions in the feature space via clustering, for
+    each region center k-nearest neighbors are used to obtain the LDA scalings,
+    which correspond to the locally discriminative basis. Currently this is
+    done at two scales `k={10,20}` if `n_feature < 50` or else `k={20,50}`.
+    """
+
+    labels, class_count = np.unique(y, return_counts=True)
+    n_class = len(labels)
+
+    n_features = X.shape[1]
+    # Number of basis yielded from each LDA
+    num_eig = min(n_class-1, n_features)
+
+    if self.n_basis is None:
+      # TODO: Get a good default n_basis directive
+      n_basis = min(20*n_features, X.shape[0]*2*num_eig)
+      warnings.warn('As no value for `n_basis` was selected, the number of '
+                    'basis will be set to n_basis= %d' % n_basis)
+
+    elif isinstance(self.n_basis, int):
+      n_basis = self.n_basis
+    else:
+      raise ValueError("n_basis should be an integer, instead it is of type %s"
+                       % type(self.n_basis))
+
+    if n_basis <= n_class:
+      raise ValueError("The number of basis should be greater than the"
+                       " number of classes")
+    elif n_basis >= X.shape[0]*2*num_eig:
+      raise ValueError("The selected number of basis needs a greater number of"
+                       " clusters than the number of available samples")
+
+    # Number of clusters needed for 2 scales given the number of basis
+    # yielded by every LDA
+    n_clusters = int(np.ceil(n_basis/(2 * num_eig)))
+
+    # TODO: maybe give acces to Kmeans jobs for faster computation?
```

Comment:
Probably not worth it. The expert user can set the backend context so that several cores are used, see
https://scikit-learn.org/stable/glossary.html#term-n-jobs

Reply:
Ok, great!